### DISCUSSION TITLE
Cache a Docker image built in workflow

### DISCUSSION THREAD

**Discussion by nergnezor** [community/community#24999] [Actions]
<p>We’re currently building a Docker image every workflow run. This takes 10-15 minutes. It’s faster to fetch the same image from Docker hub, but nicer to only be dependent on the Dockerfile in the repo. Is there any way to speed up subsequent builds by doing some kind of Docker build caching?</p>

**Comment by ethomson** [comment, marked as answer]
<p>We’re adding caching.  It will arrive by mid-November.</p>
<p>See also <a href="https://github.community/t/caching-files-between-github-action-executions/16047/13">https://github.community/t5/GitHub-Actions/Caching-files-between-GitHub-Action-executions/m-p/30974#M630</a></p>

**Comment by whoan** [comment]
<p>For those who found this Subject looking for a way to use cache in Workflows, you can use <a href="https://github.com/marketplace/actions/build-docker-images-using-cache" rel="nofollow noopener">this action</a> that does just what <a class="mention" href="/nergnezor">@nergnezor</a> mentioned: fetches the layers from a Registry and use (multiple) –cache-from to hint docker where to look for cache.</p>
<p>The action supports multi-stage builds as it pushes each stage to a registry, and pull them when a new build is triggered.</p>
<p>This is a minimal example:</p>
<pre><code>- uses: whoan/docker-build-with-cache-action@v1
  with:
    username: whoan
    password: "${{ secrets.DOCKER_PASSWORD }}"
    image_name: whoan/node
</code></pre>

**Comment by crazy-max** [comment]
<p>Using buildx and <a href="https://github.com/crazy-max/ghaction-docker-buildx" rel="nofollow noopener">ghaction-docker-buildx</a> does not require to set cache or to push to registry:</p>
<pre><code>name:buildxon:
pull\_request:
branches:masterpush:
branches:mastertags:

jobs:
buildx:
runs-on:ubuntu-lateststeps:
      -
name:Checkoutuses:actions/checkout@v1 -
name:Set up Docker Buildxid:buildxuses:crazy-max/ghaction-docker-buildx@masterwith:
version:latest -
name:Available platformsrun:echo ${{ steps.buildx.outputs.platforms }} -
name:Run Buildxrun:| docker buildx build \ --platform linux/386,linux/amd64,linux/arm/v6,linux/arm/v7,linux/arm64,linux/ppc64le,linux/s390x \ --output "type=image,push=false" \ --file ./test/Dockerfile ./test
</code></pre>
<p>Live example: <a href="https://github.com/crazy-max/swarm-cronjob/runs/231926567" rel="nofollow noopener">https://github.com/crazy-max/swarm-cronjob/runs/231926567</a></p>

**Comment by vovimayhem** [comment]
<p><a class="mention" href="/ethomson">@ethomson</a> Given <a href="https://github.com/actions/cache" rel="nofollow noopener">https://github.com/actions/cache</a> is on, does it apply to the original question? (Caching docker layers)</p>
<p>Or is something different is coming for Docker image layer caching?</p>

**Comment by kshahzada** [comment]
<p>Also looking for an answer to this; it is mid-November</p>

**Comment by duanshiqiang** [comment]
<p>It’s already mid of Feb 2020, any update on this?</p>

**Comment by n1ru4l** [comment]
<p>How is it cached then?</p>

**Comment by dimberman** [comment]
<p>Hey everyone, I found a solution that works pretty well. Unfortunately doesn’t FULLY cache, so you have to do some data transfer but seems to get the job done.</p>
<p>I run <code>docker save</code> to create a compressed image, save it as an artifact, and then load it in the next step. Not sure how this will scale but it will work for some use-cases. Still would be much better to have the ability to keep the same docker registry between steps</p>
<pre><code>jobs:
create-docker-image:
runs-on: ubuntu-latest
steps:
-uses: actions/checkout@master
-name: build image
run: docker build -t foo:bar .
-name: save image
run: docker save -o ./image foo:bar
-name: Upload image
uses: actions/upload-artifact@v1
with:
name: airflow-image
path: image
pull-image:
needs: create-docker-image
runs-on: ubuntu-latest
steps:
-name: Download image
uses: actions/download-artifact@v1
with:
name: airflow-image
-name: load image
run: docker load -i ./airflow-image/image
-name: run image
run: docker run foo:bar
</code></pre>

**Comment by pchickey** [comment]
<p>Its mid-May 2020. This still doesn’t exist and no news has been shared about when it might.</p>

**Comment by malimccalla** [comment]
<p><a class="mention" href="/ethomson">@ethomson</a> Hi, im wondering if maybe theres an update on this yet? Built in docker layer caching would be great</p>

**Comment by crazy-max** [comment]
<p>See <a href="https://github.com/crazy-max/ghaction-docker-buildx#leverage-buildx-cache" rel="nofollow noopener">https://github.com/crazy-max/ghaction-docker-buildx#leverage-buildx-cache</a></p>

**Comment by satackey** [comment]
<p>Try <a href="https://github.com/marketplace/actions/docker-layer-caching" rel="nofollow noopener">this action</a> I created.</p>

**Comment by greatvovan** [comment]
<p>How exactly to use it with Docker layers? Should I configure a step to cache/restore the dir  <code>/var/lib/docker/overlay2</code>? I prefer a general solution, without custom actions. Thanks.</p>
